### ISSUE TITLE
Build a back-end mechanism to save the off-chain roots

### ISSUE THREAD

**Issue by cedoor** [semaphore-protocol/zk-groups#34]

**Comment by jdhyun09** [comment]
**Thinking about an efficient way**
When `createGroup()` is called, it will not save the merkle root to on-chain.(because there are no members at first.)

When `addMember()` is called, I will use scheduling 'Timeout' instead of 'cron' to save off-chain roots.

Once called, set the 'Timeout' to execute the task after 1 minute. (I think we need to find the right time at the production level.)
`addMember()` call is possible while 'timeout' is set and before task is executed.(Skip set 'timeout' if 'timeout' is already on schedule)

Stores off-chain roots of addedMember groups when executing tasks.
(@todo: Find an efficient way to find updated group root through cachedGroups)

**Suggestion**
After checking the off-chain group update event, how about updating group DB members at this time?
(Do not update the DB at `addMember()` and reflect it in the DB when the root is stored on-chain.)

**Comment by jdhyun09** [comment]
> It makes sense! From what I understood, using that timeout within the addMember function should allow you to update the roots only when there are actually roots to be saved right?

yes right! It's like waking up when an actual roots change should happen. We can also accommodate many requests in a short period of time without wasting gas.  Only 1 minute but lol😂.  However, it cannot keep users waiting for a long time.

> I think the most efficient way is to mark the roots to be saved in a temporary key-value array (key: group id, which can be the hash of the group name, value: root hash), and then save just those roots.

Makes sense! Temporary array would be good way. I will try this. Thanks!

> What would be the advantage? I feel like it could become more complex for no specific reason.

I thought it would make sense to update the db when an off-chain root update actually occurs on-chain.  But db is only used to show group information, and it seems that there is no advantage because the merkle proof generation for generating the actual semaphore proof will use cachedGroup. I agree. I will withdraw this!

**Comment by cedoor** [comment]
> I thought it would make sense to update the db when an off-chain root update actually occurs on-chain. But db is only used to show group information, and it seems that there is no advantage because the merkle proof generation for generating the actual semaphore proof will use cachedGroup. I agree. I will withdraw this!

Yea, I'd say there is also another important point. Devs can decide to verify proofs off-chain, without using any contract. Those devs shouldn't wait for on-chain stuff.
